Show requested version in Solution PMUI

martinrrm opened the pull request:
## Bug

Fixes: [Show requested version in the Solution Level UI](https://github.com/nuget/home/issues/9827)
Regression: No

## Fix

Details:

@donnie-msft comments:

- I've refactored this a bit so that the GridViewColumn is dynamically added, rather than using a 0-width. I feel this is better for code readability, perf, and it does work better with Narrator (ie, it indicates correct # columns). 
  - Creating column when DataContext is changed.
- Simplified logic since `IsRequestedVisible` only needs to be set once, at PMUI creation time.
- Sorting was a little off due to the way user settings get restored. Explained in comments that I'm always creating a column header for Requested "Version", that way if the previous sort setting was this column, it can be restored during control initialization.
  - (well, actually, I found this doesn't work quite right for any column other than Project Name (https://github.com/NuGet/Home/issues/10054) .... but at least I've hooked it up correctly on this new column)
-  Renamed string to ColumnHeader_Requested to be more specific

@martinrrm comments:
Add a "Requested" column in the Solution PMUI

![image](https://user-images.githubusercontent.com/43253759/90802723-832a1380-e2cc-11ea-8101-09099508f578.png)
![image](https://user-images.githubusercontent.com/43253759/90802752-8de4a880-e2cc-11ea-89e7-8f1679855d78.png)
![image](https://user-images.githubusercontent.com/43253759/90802832-abb20d80-e2cc-11ea-96b6-66777654735b.png)
![image](https://user-images.githubusercontent.com/43253759/90803061-08adc380-e2cd-11ea-9515-c6c22087ed46.png)


## Testing/Validation

Tests Added: No  
Reason for not adding tests:  [@donnie-msft]'s thought is this is mostly a plumbing of a property to the UI. The UI is simply creating a column when necessary. The effort doesn't seem valuable given the minimal impact of the change. Open to counterpoints.

Validation:  Open Solution PMUI, with a PC and PR project. Ensure the column shows/hides as expected.
Open Project PMUI... make sure not broken.

zivkan commented on src/NuGet.Core/NuGet.PackageManagement/PublicAPI.Shipped.txt:
```diff
@@ -362,7 +362,7 @@ NuGet.ProjectManagement.BuildIntegratedInstallationContext.SuppressParent.get ->
 NuGet.ProjectManagement.BuildIntegratedInstallationContext.UnsuccessfulFrameworks.get -> System.Collections.Generic.IEnumerable<NuGet.Frameworks.NuGetFramework>
 NuGet.ProjectManagement.BuildIntegratedPackageReference
-NuGet.ProjectManagement.BuildIntegratedPackageReference.BuildIntegratedPackageReference(NuGet.LibraryModel.LibraryDependency dependency, NuGet.Frameworks.NuGetFramework projectFramework) -> void
+NuGet.ProjectManagement.BuildIntegratedPackageReference.BuildIntegratedPackageReference(NuGet.LibraryModel.LibraryDependency dependency, NuGet.Frameworks.NuGetFramework projectFramework, NuGet.Packaging.Core.PackageIdentity installedVersion) -> void
```
Granted this is in the only assembly we're more liberal in allowing breaking changes. But in general, you should not be changing shipped APIs. Instead, create a new overload, and change the existing method to call the new overload with an appropriate default value for the new parameter.

Any customers using our package in their own app and using this API will be broken if this change ships as-is.

martinrrm committed:
Display Version and Installed Columns for packages in PMUI

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/BooleanToWidthConverter.cs:
```diff
@@ -0,0 +1,26 @@
+using System;
```
license

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs:
```diff
@@ -94,6 +114,8 @@ private void UpdateInstalledVersions()
                 }
             }
 
+            IsRequestedVisible = _projects.Any(p => p.NuGetProject.ProjectStyle.Equals(NuGet.ProjectModel.ProjectStyle.PackageReference));
```
Could this boolean be set inside your loop when setting RequestedVersion? 
Or is it possible to have a `null` installedVersion but that project still should show Requested versions?

martinrrm replied:
It should show the Requested versions column, this column depends on the project type, but I can move this inside the loop

donnie-msft replied:
I suggested another refactoring. I meant that your loop already does this work, so not sure we need to call `.Any` again?

donnie-msft commented on src/NuGet.Core/NuGet.PackageManagement/PublicAPI.Unshipped.txt:
```diff
@@ -4,4 +4,4 @@ const NuGet.ProjectManagement.ProjectBuildProperties.TargetFrameworkIdentifier =
 const NuGet.ProjectManagement.ProjectBuildProperties.TargetPlatformMoniker = "TargetPlatformMoniker" -> string
-static NuGet.ProjectManagement.MessageLevelExtensions.ToLogLevel(this NuGet.ProjectManagement.MessageLevel messageLevel) -> NuGet.Common.LogLevel
\ No newline at end of file
+static NuGet.ProjectManagement.MessageLevelExtensions.ToLogLevel(this NuGet.ProjectManagement.MessageLevel messageLevel) -> NuGet.Common.LogLevel
```
this is a NoOp?

martinrrm replied:
This is not mine, I think I saved this file while rebasing and it added a newline at the end

nkolev92 replied:
If that's the case, I think you should revert this. 

Sample command: 

> git checkout dev -- src/NuGet.Core/NuGet.PackageManagement/PublicAPI.Unshipped.txt 

Should revert only that specific file.

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/SolutionView.xaml.cs:
```diff
@@ -215,5 +216,11 @@ private void SortableColumnHeader_PreviewKeyUp(object sender, KeyEventArgs e)
             }
         }
+
+        private void RequestedVersionColumnHeader_Loaded(object sender, RoutedEventArgs e)
```
Rename to Unloaded

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/SolutionView.xaml.cs:
```diff
@@ -215,5 +216,11 @@ private void SortableColumnHeader_PreviewKeyUp(object sender, KeyEventArgs e)
+        private void RequestedVersionColumnHeader_Loaded(object sender, RoutedEventArgs e)
+        {
+            var model = DataContext as PackageSolutionDetailControlModel;
+            _versionColumnHeader.Width = model.IsRequestedVisible ? _versionColumnHeader.Width : 0;
```
Awesome! So after this event, when the PMUI reappears, the width looks good? And if you expand/reduce other column widths, nothing weird happens?

martinrrm replied:
Yeah it works, but I erased the wrong function and pushed haha, it should be only `_versionColumn.Width` because in the header we are only changing the Visibility property

martinrrm committed:
Pr comments

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs:
```diff
@@ -81,8 +81,13 @@ private void UpdateInstalledVersions()
         {
             var hash = new HashSet<NuGetVersion>();
 
+            IsRequestedVisible = false;
```
I'd use `_isRequestedVisible ` here so that we don't fire PropertyChanged unnecessarily.

donnie-msft replied:
Better yet... You could use a local like `bool anyRequested = false`, then set it once down here https://github.com/NuGet/NuGet.Client/pull/3555/commits/bfd4f346c9c377deeab4ac8b2098aa6aea538855#diff-17f11a93b3135e66bd31d0d723de815aR122 (eg, `IsRequestedVisible = anyRequested`)

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs:
```diff
@@ -81,8 +81,13 @@ private void UpdateInstalledVersions()
+            IsRequestedVisible = false;
             foreach (var project in _projects)
             {
+                if (IsRequestedVisible.Equals(false) && project.NuGetProject.ProjectStyle.Equals(NuGet.ProjectModel.ProjectStyle.PackageReference))
```
In my earlier suggestion, I was thinking to simply set IsRequestedVisible inside of the `if` statement you already have here: https://github.com/NuGet/NuGet.Client/pull/3555/commits/bfd4f346c9c377deeab4ac8b2098aa6aea538855#diff-17f11a93b3135e66bd31d0d723de815aR99

martinrrm replied:
The link is not taking to a specific line of code, but I think you are referring to the if below where I check the project type, that is inside a try so it can fail

martinrrm replied:
I'm going to add an image to be more explicit, but you can have the requested column even though you don't have package with a requested version or installed, this column is going to show if there is a PR project so we don't confuse users on when this column is being show. @nkolev92

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs:
```diff
@@ -81,8 +81,13 @@ private void UpdateInstalledVersions()
+            IsRequestedVisible = false;
             foreach (var project in _projects)
             {
+                if (IsRequestedVisible.Equals(false) && project.NuGetProject.ProjectStyle.Equals(NuGet.ProjectModel.ProjectStyle.PackageReference))
```
I'd use either `!IsRequested` or `== false` since `.Equals(false)` feels a bit weird.

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs:
```diff
@@ -81,8 +81,13 @@ private void UpdateInstalledVersions()
             {
+                if (IsRequestedVisible.Equals(false) && project.NuGetProject.ProjectStyle.Equals(NuGet.ProjectModel.ProjectStyle.PackageReference))
+                {
+                    IsRequestedVisible = true;
```
use local here if you agreed with my suggestion

nkolev92 replied:
Was the decision to display the request column whenever a PR project is available? 
I thought it was only when *all* projects are PR.

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs:
```diff
@@ -76,9 +96,14 @@ private void UpdateInstalledVersions()
                         project.InstalledVersion = installedVersion.PackageIdentity.Version;
                         hash.Add(installedVersion.PackageIdentity.Version);
                         project.AutoReferenced = (installedVersion as BuildIntegratedPackageReference)?.Dependency?.AutoReferenced == true;
+                        if (project.NuGetProject.ProjectStyle.Equals(NuGet.ProjectModel.ProjectStyle.PackageReference))
```
use `using NuGet.ProjectModel`.
Don't fully qualify the enum

nkolev92 replied:
I added a PR to improve our guidelines with this: https://github.com/NuGet/NuGet.Client/pull/3592

nkolev92 replied:
You are already doing a project style check above right?

donnie-msft replied:
I think you're asking what I asked here https://github.com/NuGet/NuGet.Client/pull/3555#discussion_r473109979
and now here https://github.com/NuGet/NuGet.Client/pull/3555#discussion_r473268210

martinrrm replied:
Yes, but this is for the package, here we are comparing if the package installed is from a PR project, otherwise the package doesn't have a requested version, because PC project don't use Allowed Version

nkolev92 reviewed (commented):
Can you please add screenshots for the relevant scenarios? 

personally I'd like to see example for: 

* PR project level.
* PR only solution
* PC only solution
* PR + PC mixed solution

martinrrm committed:
Pr comments

Fixed a bug for the column, there is times that the component is lot being loaded/unloaded correctly, so we need to use both properties

martinrrm committed:
File conflict

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/BooleanToWidthConverter.cs:
```diff
@@ -0,0 +1,29 @@
+        {
+            var boolValue = (bool)value;
+
+            return boolValue ? "Auto" : "0";
```
Have you had a chance to test the narrator here? 
https://github.com/NuGet/NuGet.Client/blob/dev/docs/feature-guide.md#visual-studio-ui-considerations

martinrrm replied:
Yeah it is getting announced when the width is 0, the narrator says 4 columns

donnie-msft replied:
Hmm... would we want to allow screen readers to indicate the column is blank? Or should the column be logically removed?

donnie-msft replied:
I completely refactored this and Narrator indicates correct # columns now.

nkolev92 reviewed (dismissed):
I think you should test the narrator to make sure the 0-width column does not get announced.
Looks great beyond that. 

Only thing I'm concerned is that the 0 width column might get announced.

donnie-msft committed:
Revert PublicAPI.Unshipped.txt

donnie-msft committed:
resolve merge conflict

donnie-msft committed:
don't fully qualify enum

donnie-msft committed:
Requested column is now dynamically created. Reduced logic for IsRequestedVisible. Tweaked width for sorting icon

donnie-msft committed:
Refactor so Requested Header sorting can be restored from usersettings

donnie-msft committed:
Revert BooleanToWidthConverter

donnie-msft committed:
tweak column widths

donnie-msft commented:
@rrelyea please take a look at description for my comments on this PR.
@nkolev92 dismissed your review as I've made a lot of changes here.

donnie-msft committed:
fix formatting

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs:
```diff
@@ -129,9 +133,15 @@ private async Task UpdateInstalledVersionsAsync(CancellationToken cancellationTo
                         hash.Add(installedVersion.Identity.Version);
                         project.AutoReferenced = installedVersion.IsAutoReferenced;
+
+                        if (project.NuGetProject.ProjectStyle.Equals(ProjectStyle.PackageReference))
```
How about `if (IsRequestedVisible)` ? 

```suggestion
                        if (IsRequestedVisible)
```

donnie-msft replied:
Can't some projects be PR and others be PC? I thought this was project-by-project determination

nkolev92 replied:
You are right, it was a different method from the above one.
I'm pretty sure I provided the same feedback to Martin :) 

Resolving this.

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/SolutionView.xaml.cs:
```diff
@@ -49,15 +50,36 @@ public SolutionView()
+        {
+            //Since this event will fire when closing PMUI, when the model is set to Project PMUI,
+            //and when other WPF controls in the tree set their DataContext, this cast is more important than it may seem.
+            var model = e.NewValue as PackageSolutionDetailControlModel;
```
No need to change it, but `is` has an optimization so you can write something like:

```cs
if ( e.NewValue is PackageSolutionDetailControlModel model)
```

donnie-msft replied:
Did it since was making other changes

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/SolutionView.xaml.cs:
```diff
@@ -226,5 +248,29 @@ private void SortableColumnHeader_PreviewKeyUp(object sender, KeyEventArgs e)
             }
         }
+
+        private GridViewColumn CreateRequestedVersionColumn()
```
I find this weird, but sounds like it solves the narrator problem. 
I'd wait on a review from someone skilled with xaml stuff.

donnie-msft replied:
@rrelyea & @srdjanjovcic

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/SolutionView.xaml.cs:
```diff
@@ -197,10 +219,10 @@ private void ListView_SizeChanged(object sender, SizeChangedEventArgs e)
         {
             // GridViewColumnHeader doesn't handle setting minwidth very well so we prevent it here.
-            if (sizeChangedEventArgs.NewSize.Width <= 60)
+            if (sizeChangedEventArgs.NewSize.Width <= 67)
```
how did we come up with a diff of 7? :D

donnie-msft replied:
It's arbitrary, but the word Installed is a tiny bit wider than Version, so that's why I expanded

donnie-msft replied:
You inspired me to make this vary based on column name. I also shrunk to 64 since it still looks good at that spacing.

nkolev92 reviewed (approved):
Looks fine overall to me. 

Couple of small questions. 
I'd prefer if someone else approved the requested column addition trick. 
I am happy that it solves the problem, and don't have any particular ideas, but it increases the code complexity in a way that's not obvious.

donnie-msft committed:
Header widths now based on column name

donnie-msft committed:
model casting optimization

donnie-msft committed:
Merge branch 'dev' into dev-martinrrm-RequestedVersionInSolutionUI
